source/develop/release-management/features/storage/multipath-events.html.md

```diff
@@ -0,0 +1,167 @@
+- A topic monitor will be created with the first listener.
+- Once the monitor receives an event, it will call the listener callback
+(in a different thread than the one the listener registered)
+- A listener can unregister from a topic.
+- The topic monitor will stop monitoring once the last listener unregister.
+
+#### Udev event monitor
+
+A Udev event monitor is introduced, listening to device-mapper udev events.
+The relevant events for this feature are PATH_FAILED and PATH_REINSTATED:
+
+Variable Name: DM_ACTION
+Uevent Action(s): KOBJ_CHANGE
+Type: string
+Description:
+Value: Device-mapper specific action that caused the uevent action.
+	PATH_FAILED - A path has failed.
+	PATH_REINSTATED - A path has been reinstated.
+
+- The Vdsm will register a listener to the Udev event monitor when connecting
+to the storage pool.
+- The listener will send an event to the engine when its callback is called.
+- The Vdsm will unregister the listener to the Udev event monitor when 
+disconnecting to the storage pool.
+- The user can disable this feature in a specific host in Vdsm configuration file
+
+### New engine API
+
+New audit logs are added:
+- Multipath device path up
+- Multipath device path down
+
+## Benefit to oVirt
+
+- The user will be able to detect storage issues faster.
+
+
+## User Experience
+
+New audit logs are added with information on the device path status.
+
+
+## Installation/Upgrade
+
+This feature does not affect engine or Vdsm installation or upgrades
+
+
+## User work-flows
+
+Once a user connects to an ISCSI server, 'Multipath device path up' event
+will be created.
+If there is an issue with the connection to the storage server, and the 
+multipath reports an issue with one or more device path,
+'Multipath device path down' will be created.
+Once the issue is fixed, 'Multipath device path up' event will be created.
+
+
+## Dependencies / Related Features
+
+Additional package dependency in Vdsm:
+- pyudev
+
+No related features.
+
+
+### Entity Description
+
+NA
+
+## CRUD
+
+NA
+
+
+## Event Reporting
+
+- 'Multipath device path up'
+- 'Multipath device path down'
+
+
+## Documentation / External references
+
+This feature is required for resolving
+[Bug 723931](https://bugzilla.redhat.com/723931)
+
+Pyudev - [Monitoring devices](http://pyudev.readthedocs.io/en/latest/guide.html#monitoring-devices)
+
+Uevent [example](https://www.kernel.org/doc/Documentation/device-mapper/dm-uevent.txt)
+
+## Testing
+
+- Connect to a new ISCSI storage server
+  - For each multipath device path a 'Multipath device path up' is triggered
+
+
+### Negative flows
+
+- Block network to ISCSI storage server
+  - For each multipath device path a 'Multipath device path down' is triggered
+
+
+## Release Notes
+
+TBD
+
+## Open Issues
+
+- Connecting to a new ISCSI storage server will trigger a lot of events,
+that are not useful to the user.
+- In a scale environment, issue on the storage server or network can trigger
+an event storm.
```

Comment:
Maybe we should ignore messages for a device if they were already reported on other hosts.

For example:

1. Device X on host A path went down, 3 paths active of 4 - log
2. Device X on host B path went down, 3 paths active of 4 - ignore (already reported)
3. Device X on host C path went down, 0 paths active of 4 - log - device cannot be used now 
4. Device X on host A path went down, 0 paths active of 4 - log - ignore (already reported)

We can keep a memory mapping of all devices, report changes when the event changes the mapping.

I think we have several priorities of events:
- info: number of valid paths increased on some host
- warnings: number of valid path decrease, but the device is still usable
- error: number valid paths is 0 on some host - the device is not usable or will be unusable soon (depending on multipath configuration)

We assume that all host must see all paths of all devices, so we can focus on the device state instead of the host state. We are already handling host state via storage monitoring.

Reply:
I'm not sure the assumption that all hosts must see all paths is a valid assumption. They may be connected differently, via different number of initiators or to different number of targets. (In fact, this is a good 'soft' warning to the user - which he may be OK with or act upon if it wasn't intended) - but this is a bit too complex for this implementation ATM.
I would also not drop any alert - imagine you've put one switch down for maintenance, which affected (as you anticipated, host A. You really do wish to know if anything bad happened to host B, which was on a different switch that was not supposed to go down!

Reply:
This kind of logic can be misleading. You can think that host B is OK by ignoring it.
Let's start with keeping it simple and see how it is a real issue.

Reply:
We can start simple for development, but I don't think this feature will be usable if we don't have a way to aggregate and filter events.